libcalico-go/lib/dispatcher/blocking.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dispatcher
+
+import (
+	"context"
+	"errors"
+
+	"github.com/sirupsen/logrus"
+)
+
+// BlockingDispatcher blocks on an input channel until stop,
+// and broadcasts each input message to all output channels.
+// Blocks until all output channels have consumed a message,
+// before pulling in the next input message.
+type BlockingDispatcher[T any] struct {
+	outputs []chan<- T
+	input   <-chan T
+
+	log *logrus.Entry
+}
+
+// NewBlockingDispatcher returns a BlockingDispatcher which
+// will consume messages from the input channel.
+// Dispatcher must be started with Start(ctx). Stopped by cancelling ctx.
+func NewBlockingDispatcher[T any](input <-chan T) (*BlockingDispatcher[T], error) {
+	log := logrus.WithField("component", "BlockingDispatcher")
+
+	if input == nil {
+		return nil, errors.New("No input channel provided for dispatcher to consume")
```

Comment:
I'd usually panic rather than returning an error in this sort of case.  It's a programming bug to forget to pass a valid argument so the caller likely won't be able to handle it (since they caused the problem!)

Handling the extra error is just a bit tedious.